### ISSUE TITLE
'rye sync' + uv resolves the incorrect version w.r.t platform requirements (e.g python_version < '3.11')

### ISSUE THREAD

**Issue by mvaled** [astral-sh/rye#803]
### Steps to Reproduce

Create a simple repository

```
rye init test1
```

Edit the `pyproject.toml` to add dependencies:

```
dependencies = [
    "xotl.tools~=2.2.5; python_version <= '3.11'",
    "xotl.tools>=3.0.0a5; python_version > '3.11'",
]
```

Pin Python 3.11 and sync

```
rye config --set-bool behavior.use-uv=true
rye pin --relaxed 3.11
rye sync
Initializing new virtualenv in /home/manu/tmp/test1/.venv
Python version: cpython@3.11.7
Generating production lockfile: /home/manu/tmp/test1/requirements.lock
Generating dev lockfile: /home/manu/tmp/test1/requirements-dev.lock
Installing dependencies
   Built file:///home/manu/tmp/test1                                                                                                                                                          Built 1 editable in 234ms
Installed 4 packages in 2ms
 + packaging==23.2
 + test1==0.1.0 (from file:///home/manu/tmp/test1)
 + typing-extensions==4.10.0
 + xotl-tools==3.0.0a5
```

Notice the chosen version of `xotl.tools` is incorrect for Python 3.11.

### Expected Result

The selected version of the package should be the one matching the platform requirements.

### Actual Result

It's selecting either by the platform that runs rye, or it's selecting the most recent version of the package.

### Version Info

rye 0.27.0
commit: 0.27.0 (43ee4fce0 2024-02-26)
platform: linux (x86_64)
self-python: cpython@3.12
symlink support: true
uv enabled: true

### Notes

When disabling `use-uv`, the selected version of the package is correct:

```
rye config --set-bool behavior.use-uv=false
rye sync
Reusing already existing virtualenv
Generating production lockfile: /home/manu/tmp/test1/requirements.lock
Creating virtualenv for pip-tools
Generating dev lockfile: /home/manu/tmp/test1/requirements-dev.lock
Installing dependencies
   ...
Successfully installed test1-0.1.0 xotl-tools-2.2.5
```

```
rye pin --relaxed 3.12
rye sync
Python version mismatch (found cpython@3.11.7, expected cpython@3.12.1), recreating.
Initializing new virtualenv in /home/manu/tmp/test1/.venv
Python version: cpython@3.12.1
Generating production lockfile: /home/manu/tmp/test1/requirements.lock
Generating dev lockfile: /home/manu/tmp/test1/requirements-dev.lock
Installing dependencies
   ...
Successfully installed packaging-23.2 test1-0.1.0 typing-extensions-4.10.0 xotl-tools-3.0.0a5
Done!
```

**Comment by mvaled** [comment]
Notice that using `uv pip install -r requirements.txt` directly works.  The contents of the file `requirements.txt` being:

```
xotl.tools~=2.2.5; python_version <= '3.11'
xotl.tools>=3.0.0a5; python_version > '3.11'
```

The output of the command shows that the correct version is selected:

```
$ source .venv/bin/active
$ uv pip install -r requirements.txt                                                                                                                                                                Resolved 2 packages in 549ms
Downloaded 1 package in 108ms
Installed 1 package in 5ms
 - xotl-tools==3.1.0a1
 + xotl-tools==2.2.5
```

So, I'm guessing the bug is that `rye` is calling `uv` in a way that it uses rye's bootstrapping python version.

**Comment by ischaojie** [comment]
It looks like the latest version fixed this:

```
➜  test rye init test1
success: Initialized project in /Users/zz/projects/test/test1
  Run `rye sync` to get started
➜  test cd test1
➜  test1 git:(main) ✗ vim pyproject.toml
➜  test1 git:(main) ✗ rye config --set-bool behavior.use-uv=true
➜  test1 git:(main) ✗ rye pin --relaxed 3.11
pinned 3.11 in /Users/zz/projects/test/test1/.python-version
➜  test1 git:(main) ✗ rye sync
Initializing new virtualenv in /Users/zz/projects/test/test1/.venv
Python version: cpython@3.11.5
Generating production lockfile: /Users/zz/projects/test/test1/requirements.lock
Generating dev lockfile: /Users/zz/projects/test/test1/requirements-dev.lock
Installing dependencies
   Built file:///Users/zz/projects/test/test1                                                                                                                              Built 1 editable in 158ms
Resolved 1 package in 0.58ms
Downloaded 1 package in 486ms
Installed 3 packages in 6ms
 + test1==0.1.0 (from file:///Users/zz/projects/test/test1)
 + typing-extensions==4.10.0
 + xotl-tools==2.2.5
Done!
```

**Comment by mvaled** [comment]
@ischaojie 

Yes! It has been solved.
